hack/migrate-etcd

```diff
@@ -0,0 +1,200 @@
+#!/bin/bash
+set -u
+function getNamespaces()
+{
+    namespaces=`kubectl get ns --kubeconfig=${gardenkubeconfig} -o custom-columns=NAME:.metadata.name | tail -n +2`
+    echo "${namespaces[@]}"
+}
+
+function getShootsInNamespace()
+{
+    local namespace=$1
+    local shoots=`kubectl get shoot --kubeconfig=${gardenkubeconfig} -n ${namespace} -o custom-columns=NAME:.metadata.name | tail -n +2`
+    echo "${shoots[@]}"
+}
+
+function processShoot()
+{
+    local shoot=$1
+    local namespace=$2
+    local shootJson=`kubectl get shoot ${shoot} --kubeconfig=${gardenkubeconfig} -n ${namespace} -o json`
+    if [[ "$?" != "0" ]] || [[ "${shootJson}" == "" ]];
+    then 
+        echo "Shoot information not available for ${namespace}/${shoot}."
+        return 
+    fi
+    local hibernation_status=`echo -n ${shootJson} | jq .spec.hibernation.enabled`
+    if [ "$hibernation_status" != "true" ]
+    then
+        echo "${namespace}/${shoot} is not hibernated. Skipping forced migration."
+        return
+    fi
+    
+    if [ "${skip_reconcile_check}" != "true" ]
+    then
+        local reconcile_enabled=`echo -n ${shootJson} | jq .metadata.annotations | grep 'shoot.garden.sapcloud.io/ignore' | wc -l`
+        if [ ! "${reconcile_enabled}" == "0" ]
+        then
+            echo "${namespace}/${shoot} is not reconciled. Skipping forced migration."
+            return
+        fi
+    fi
+    
+    local seed=`echo -n ${shootJson} | jq .spec.cloud.seed | sed s/\"//g`
+    fetchKubeconfigForSeed ${seed}
+    if [[ ! -e $KUBECONFIGS_DIR/seed-${seed}.kubeconfig ]]
+    then
+        echo "Failed to fetch kubeconfig for seed ${seed}"
+        return
+    fi
+    
+    namespace_in_seed=`echo -n ${shootJson} | jq .status.technicalID | sed s/\"//g`
+    has_etcd_main=`kubectl get statefulset etcd-main -n ${namespace_in_seed} --kubeconfig=$KUBECONFIGS_DIR/seed-${seed}.kubeconfig | wc -l`
+    if [ "${has_etcd_main}" == "0" ]
```

Comment:
Irrespective of whether shoot is hibernated or not, we delete the old etcd-main statefulset and create new etcd-main statefulset with updated volumnClaimTemplate. So, within this period, script might not find etcd-statefulset but it might find old PVC. So, at this stage skip the migration if etcd-statefulset is not there and old pod and pvc i.e. etcd-main-0 pod and  etcd-main-etcd-main-0 pvc is unavailable.

Reply:
This script is only meant to for forcing etcd to migrate for hibernated clusters. So in case of non-hibernated clusters we should not depend on this script at all. I am checking if `etcd-main` statefulset is there. If not, I return. Should I handle it differently?

Reply:
The above mentioned scenario is applicable for hibernated cluster. Hence add the missing checks for PVC. Current script wont migrate some of the hibernated clusters which are under first reconciliation post enabling hibernation at the time of application of script.